Compiling locally for iOS gives a java.lang.OutOfMemoryError: Java heap space

With 7.0.256, when I try to compile locally for iOS, I get the error below. The compilation works OK for building the Simulator locally. 

It seems to be parsing literally every class in CN1 (including the many recent new ones) - not sure if it did so before, but at least I never noticed. 

Not sure what info is needed to debug this, I've kept the full trace so let me know if I should add more details. 

```
[ERROR] iOS builder log:
Executing: /usr/bin/xcrun -f xcodebuild Executing: /usr/bin/xcodebuild -version Executing: /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/jre/../bin/javac -version Executing: /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/jre/../bin/javac -source 1.6 -target 1.6 -classpath /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes -d /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/net/informaticalibera/cn1/nativelogreader/NativeLogsReaderImplCodenameOne.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/net/informaticalibera/cn1/nativelogreader/NativeLogsReaderStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/org/littlemonkey/connectivity/NativeConnectivityImplCodenameOne.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/org/littlemonkey/connectivity/NativeConnectivityStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/myappStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/myapp/calendar2/impl/CalendarNativeInterfaceImplCodenameOne.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/myapp/calendar2/impl/CalendarNativeInterfaceStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/diamonddevgroup/device/DeviceNativeImplCodenameOne.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/diamonddevgroup/device/DeviceNativeStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/codename1/nui/NTextFieldNativeStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/codename1/nui/NSelectNativeStub.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/codename1/nui/NTextFieldNativeImplCodenameOne.java /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/stub/com/codename1/nui/NSelectNativeImplCodenameOne.java warning: [options] bootstrap class path not set in conjunction with -source 1.6
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
1 warning
Executing: java -DsaveUnitTests=false -DfieldNullChecks=false -DINCLUDE_NPE_CHECKS=true -Dcn1.onDeviceDebug=false -DbundleVersionNumber=0.17 -Xmx384m -jar /var/folders/b4/_k9_sbns6nb6m8r8f2snd6n80000gn/T/temp5334902631120218146.jar ios /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes;/Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/res;/Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/btres /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build myapp com.myapp.myapp myapp 0.17 ios EventKit.framework;SystemConfiguration.framework;UserNotifications.framework;BackgroundTasks.framework;WebKit.framework;CoreImage.framework;QuartzCore.framework;StoreKit.framework Root is: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/dist
srcRoot is: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/dist/myapp-src
Parsing: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes/net/informaticalibera/cn1/nativelogreader/NativeLogs.class
Parsing: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes/net/informaticalibera/cn1/nativelogreader/NativeLogsReaderStub.class
Parsing: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes/net/informaticalibera/cn1/nativelogreader/NativeLogsReader.class
Parsing: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes/net/informaticalibera/cn1/nativelogreader/NativeLogsReaderImplCodenameOne.class
Parsing: /Users/myuser/myapp/ios/target/codenameone/antProject/dist/ios-build/classes/org/littlemonkey/connectivity/Connectivity$ConnectionState.class
```


```
Native files total 2044K
Error while working with the class: com_codename1_router_PopGuard
java.lang.OutOfMemoryError: Java heap space
	at com.codename1.tools.translator.NativeSymbolIndex.newState(NativeSymbolIndex.java:73)
	at com.codename1.tools.translator.NativeSymbolIndex.extend(NativeSymbolIndex.java:78)
	at com.codename1.tools.translator.NativeSymbolIndex.<init>(NativeSymbolIndex.java:66)
	at com.codename1.tools.translator.Parser.getNativeSymbolIndex(Parser.java:304)
	at com.codename1.tools.translator.BytecodeMethod.isMethodUsedByNative(BytecodeMethod.java:596)
	at com.codename1.tools.translator.ByteCodeClass.calcUsedByNative(ByteCodeClass.java:2321)
	at com.codename1.tools.translator.ByteCodeClass.markDependencies(ByteCodeClass.java:277)
	at com.codename1.tools.translator.Parser.writeOutput(Parser.java:638)
	at com.codename1.tools.translator.ByteCodeTranslator.handleIosOutput(ByteCodeTranslator.java:579)
	at com.codename1.tools.translator.ByteCodeTranslator.main(ByteCodeTranslator.java:204)
```

--- a/maven/codenameone-maven-plugin/src/main/java/com/codename1/builders/IPhoneBuilder.java
+++ b/maven/codenameone-maven-plugin/src/main/java/com/codename1/builders/IPhoneBuilder.java
@@ -2682,7 +2682,33 @@ public class IPhoneBuilder extends Executor {
                     // both the iOS app target and the tvOS target.
                     parparCmd.add(tvNativeBuilder.parparvmOptionalFrameworksArg());
                 }
-                parparCmd.add("-Xmx384m");
+                // Pass through extra translator JVM options (notably a larger
+                // -Xmx) from the CN1_TRANSLATOR_OPTS environment variable. The
+                // forked JVM does not inherit the Maven process's -D properties,
+                // so this is the only way to reach the translator for tuning.
+                String translatorOpts = System.getenv("CN1_TRANSLATOR_OPTS");
+                boolean heapOverridden = false;
+                if (translatorOpts != null && !translatorOpts.trim().isEmpty()) {
+                    for (String opt : translatorOpts.trim().split("\\s+")) {
+                        if (!opt.isEmpty()) {
+                            parparCmd.add(opt);
+                            if (opt.startsWith("-Xmx")) {
+                                heapOverridden = true;
+                            }
+                        }
+                    }
+                }
+                // Default heap; a -Xmx in CN1_TRANSLATOR_OPTS takes precedence.
+                // The dead-code cull builds an in-memory suffix automaton over
+                // all native symbols (NativeSymbolIndex, from #5236) to avoid the
+                // old O(N^2) substring scan that timed out on large apps -- that
+                // index trades time for memory, so the historical 384m cap now
+                // OOMs local iOS builds as the CN1 class count grows (issue
+                // #5344). The cloud builder already runs the same translator at
+                // 1024m; match it here so local and server builds behave alike.
+                if (!heapOverridden) {
+                    parparCmd.add("-Xmx1024m");
+                }
                 parparCmd.add("-jar");
                 parparCmd.add(parparVMCompilerJar);
                 parparCmd.add("ios");

